src/test/shell/bazel/bazel_cc_code_coverage_test.sh

```diff
@@ -81,47 +96,54 @@ int main(void) {
 }
 
-# Reads the list of arguments provided by the caller (using $@) and uses them
-# to produco .gcno files using g++.
-function generate_gcno_files() {
-  # "-fprofile-arcs -ftest-coverage" tells the compiler to generate coverage
-  # information needed by gcov and include additional code in the object files
-  # for generating the profiling.
-  g++ -fprofile-arcs -ftest-coverage "$@" && return 0
-  fail "Couldn't produce .gcno files for $@"
-  return 1
-}
-
-# Reads the list of arguments provided by the caller (using $@) and uses them
-# to produce an instrumented binary using g++.
+# Generates and executes an instrumented binary:
+#
+# Step 1: Reads the list of arguments provided by the caller (using $@) and uses them
+# to produce an instrumented binary using g++. This step also generates
+# the notes (.gcno) files.
+#
+# Step 2: Executes the instrumented binary. This step also generates the
+# profile data files.
 # - path_to_binary destination of the binary produced by g++
-function generate_instrumented_binary() {
+function generate_and_execute_instrumented_binary() {
   local path_to_binary="${1}"; shift
-  # "-fprofile-arcs -ftest-coverage" tells the compiler to generate coverage
-  # information needed by gcov and include additional code in the object files
-  # for generating the profiling.
-  g++ -fprofile-arcs -ftest-coverage "$@" -o "$path_to_binary"  && return 0
-  fail "Couldn't produce the instrumented binary for $@ \
-      with path_to_binary $path_to_binary"
-  return 1
-}
-
-# Execute an instrumented binary and generate the gcda file.
-# - path_to_binary path of instrumented binary
-function generate_gcda_file() {
-  local path_to_binary="${1}"
-  "$path_to_binary" && return 0
-  fail "Couldn't execute the instrumented binary $path_to_binary"
-  return 1
+  local gcda_directory="${1}"; shift
+  # -fprofile-arcs   Instruments $path_to_binary. During execution the binary
+  #                  records code coverage information.
+  # -ftest-coverage  Produces a notes (.gcno) file that coverage utilities
+  #                  (e.g. gcov, lcov) can use to show a coverage report.
+  # -fprofile-dir    Sets the directory where the profile data (gcda) appears.
+  #
+  # The profile data files need to be at a specific location where the C++
+  # coverage scripts expects them to be ($COVERAGE_DIR/path/to/sources/).
+  g++ -fprofile-arcs -ftest-coverage \
+      -fprofile-dir="$gcda_directory" \
+      "$@" -o "$path_to_binary"  \
+       || fail "Couldn't produce the instrumented binary for $@ \
+            with path_to_binary $path_to_binary"
+
+   # g++ generates the notes files in the current directory. The documentation
+   # (https://gcc.gnu.org/onlinedocs/gcc/Gcov-Data-Files.html#Gcov-Data-Files)
+   # says they are placed in the same directory as the object file, but they
+   # are not. Therefore we move them in the same directory.
+   mv a.gcno coverage_srcs/a.gcno
```

Comment:
These hardcoded paths are misplaced, aren't they? They should be parameterized, if that's too hard then do the hardcoded moves outside of the function. As it stands it can only apply to the parameters you're passing above defeating the purpose of the function.

Reply:
Good point, I've moved them outside.